plugins/device/rest.go

```diff
@@ -129,3 +130,91 @@ func deviceListHandler(w http.ResponseWriter, r *http.Request) {
 	restutils.SendHTTPResponse(ctx, w, http.StatusOK, devices)
 
 }
+
+func deviceEditStateHandler(w http.ResponseWriter, r *http.Request) {
+
+	ctx := r.Context()
+	logger := gdctx.GetReqLogger(ctx)
+	peerID := mux.Vars(r)["peerid"]
+	if uuid.Parse(peerID) == nil {
+		restutils.SendHTTPError(ctx, w, http.StatusBadRequest, "Invalid peer-id passed in url")
+		return
+	}
+
+	req := new(deviceapi.EditDeviceStateReq)
+	if err := restutils.UnmarshalRequest(r, req); err != nil {
+		logger.WithError(err).Error("Failed to Unmarshal request")
+		restutils.SendHTTPError(ctx, w, http.StatusBadRequest, errors.ErrJSONParsingFailed)
+		return
+	}
+
+	var deviceState string
+	deviceState, exists := deviceapi.DeviceStates[strings.ToLower(req.State)]
+	if !exists {
+		logger.Error("State provided in request doesnot match any state supported by device")
+		errMsg := fmt.Sprintf("Invalid state, Supported states are (%v)", getKeys(deviceapi.DeviceStates))
+		restutils.SendHTTPError(ctx, w, http.StatusBadRequest, errMsg)
+		return
+	}
+
+	txn, err := transaction.NewTxnWithLocks(ctx, peerID)
+
+	if err != nil {
+		if err == transaction.ErrLockTimeout {
+			restutils.SendHTTPError(ctx, w, http.StatusConflict, err)
+		} else {
+			restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		}
+		return
+	}
+	defer txn.Done()
+
+	peerInfo, err := peer.GetPeer(peerID)
+	if err != nil {
+		logger.WithError(err).WithField("peerid", peerID).Error("Peer ID not found in store")
+		if err == errors.ErrPeerNotFound {
+			restutils.SendHTTPError(ctx, w, http.StatusNotFound, errors.ErrPeerNotFound)
+		} else {
+			restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, "Failed to get peer details from store")
+		}
+		return
+	}
+
+	var device deviceapi.Info
+	if _, exists := peerInfo.Metadata["_devices"]; !exists {
+		logger.WithError(err).WithField("peerid", peerID).Error("Peer doesnot contain any devices")
+		restutils.SendHTTPError(ctx, w, http.StatusBadRequest, "Peer doesnot contain any devices")
+	} else {
+		devices, err := deviceutils.GetDevices(peerID)
```

Comment:
I see multiple calls to GetPeer/device. Expose new function from store-utils `deviceutils.SetDeviceState(peerID, dev, state)`. Return error if device not found, peer not found. Also skip the last AddDevice call.

Also you have to make sure to update the size if you ever add new device. This API should always update existing device and not add new entry